File: 000801-001000/000921.使括号有效的最少添加/921.js
/**
 * @param {string} s
 * @return {number}
 */
// 方法一：用时 68 ms
// 执行用时超过了 88 %
// 消耗内存超过了 44 %
// 难度: 中等
var minAddToMakeValid = function(s) {
    let ret = 0
    let mstack = 0
    for (const item of s) {
        if (item == ')') {
            if (mstack > 0) {
                mstack --
                continue
            }
            ret ++
            continue
        }
        mstack ++
    }
    return (ret + mstack)
};

console.log(minAddToMakeValid("())"))
console.log(minAddToMakeValid("((("))
console.log(minAddToMakeValid("()"))
console.log(minAddToMakeValid("()))(("))

File: 000801-001000/000921.使括号有效的最少添加/921.go
package main

import "fmt"

// 方法一：用时 0 ms
// 执行用时超过了 100 %
// 消耗内存超过了 100 %
// 难度: 中等
func minAddToMakeValid(s string) int {
	ret := 0
	mstack := 0
	for i := 0; i < len(s); i++ {
		if s[i] == ')' {
			if mstack > 0 {
				mstack --
				continue
			}
			ret ++
			continue
		}
		mstack++
	}
	return ret + mstack
}

func main ()  {
	fmt.Println(minAddToMakeValid("())"))
	fmt.Println(minAddToMakeValid("((("))
	fmt.Println(minAddToMakeValid("()"))
	fmt.Println(minAddToMakeValid("()))(("))
}
